### PR TITLE
Remove setuptools_scm_git_archive dependency and add sdist test

### LINKED ISSUE #23325: [Bug]: Inconsistent version when install 3.5.0 to 3.5.2 on GitHub CI
### Bug summary

Not sure if this is an MPL issue, but here goes.

The SymPy tests cannot install the latest pypi versions of MPL as the message `filename has '3.5.2', but metadata has '0.0'` is obtained. See more below. (It ends up installing 3.4.3.)

### Code for reproduction

```python
pip install matplotlib
```

### Actual outcome

```
Collecting matplotlib
  Downloading matplotlib-3.5.2.tar.gz (35.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.2/35.2 MB 56.8 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Discarding https://files.pythonhosted.org/packages/2f/be/7d6e073a3eb740ebeba43a69f5de2b141fea50b801e24e0ae024ac94d4ac/matplotlib-3.5.2.tar.gz#sha256=48cf850ce14fa18067f2d9e0d646763681948487a8080ec0af2686468b4607a2 (from https://pypi.org/simple/matplotlib/) (requires-python:>=3.7): Requested matplotlib from https://files.pythonhosted.org/packages/2f/be/7d6e073a3eb740ebeba43a69f5de2b141fea50b801e24e0ae024ac94d4ac/matplotlib-3.5.2.tar.gz#sha256=48cf850ce14fa18067f2d9e0d646763681948487a8080ec0af2686468b4607a2 has inconsistent version: filename has '3.5.2', but metadata has '0.0'
```

### Expected outcome

metadata is not 0.0

### Additional information

_No response_

### Operating system

_No response_

### Matplotlib Version

3.5.0 to 3.5.2

### Matplotlib Backend

_No response_

### Python version

_No response_

### Jupyter version

_No response_

### Installation

_No response_

### LINKED ISSUE #23386: setuptools_scm-git-archive is obsolete
### Bug summary

setuptools_scm starting with version 7 has native support for git archives, so this dep can go away ; in fact, the [project homepage](https://github.com/Changaco/setuptools_scm_git_archive) is a migration guide!

### PR DESCRIPTION
**Pull request by oscargus**
## PR Summary
Closes #23325
Closes #23386

Removes setuptools_scm_git_archive as dependency, but increases the setuptools_scm version requirement from 4 to 7.

## PR Checklist

**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [N/A] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by timhoffm — APPROVED** [review]

**Comment by QuLogic** [comment]
In Cartopy, we removed `.git_archival.txt` entirely: SciTools/cartopy#2057; I wonder if either approach is broken somehow. @dopplershift @greglucas

I think wheels were always okay; the problem is in #23325 where installs are made from an sdist (because 3.11 has no wheels). So this needs testing from sdist, _not_ just cibuildwheel.

**Comment by dopplershift** [comment]
I think we broke installs from direct GitHub tarballs.

**Comment by oscargus** [comment]
OK! I'll try to install 3.11 on some computer and see what happens. (It wasn't clear to me that this was actually related to #23325, just acted on the info in #23386.)

**Comment by oscargus** [comment]
I tested this locally and it seems to solve #23325 as well.

Tried to add a CI-file building an sdist and printing the resulting version number. Not sure how worthwhile it is. (Clearly, this should be connected to a label and more OS and Python versions, but to illustrate it.)

**Comment by oscargus** [comment]
It seems like cibuildwheel cannot first build an sdist. Current version uses build and then installs both sdist and wheel and checks the version number.

**Comment by tacaswell** [comment]
Is this a blocking issues for 3.6?

**Comment by oscargus** [comment]
> Is this a blocking issues for 3.6?

Hard to say. I'd say that it is rather non-controversial. 
* Removing `setuptools_scm_git_archive` and modifying `.git_archival.txt` are cookbook recipes. This is primarily maintenance, although it happens to solve the sdist version 0.0 problem as well.
* Requiring `setuptools_scm` >= 7, I cannot really judge, probably the main problem is if that in turn require some "problematic" versions of something.
* Testing is probably the main question. I am not sure that we actually do require a test for the version not being "0.0" again. It would be nice, but maybe it is better to drop the new tests, rather than delaying the fix?

**Comment by QuLogic** [comment]
This definitely needs to go in 3.6, but it depends on #23557.

File: .github/workflows/cibuildsdist.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+on:
+  # Save CI by only running this on release branches or tags.
+  push:
+    branches:
+      - main
+      - v[0-9]+.[0-9]+.x
+    tags:
+      - v*
+  # Also allow running this action on PRs if requested by applying the
+  # "Run cibuildsdist" label.
```
Comment (QuLogic) [inline .github/workflows/cibuildsdist.yml:12]: Not sure we really need a separate label for this one.

Comment (tacaswell) [reply .github/workflows/cibuildsdist.yml:12]: I would use the same label we use for the cibuild wheel tests?

Comment (oscargus) [reply .github/workflows/cibuildsdist.yml]: Wasn't sure, but have modified it to use the same label.

File: .git_archival.txt [hunk 2]
```diff
@@ -1 +1,4 @@
+node: $Format:%H$
+node-date: $Format:%cI$
+describe-name: $Format:%(describe:tags=true)$
 ref-names: $Format:%D$
```
Comment (tacaswell) [inline .git_archival.txt:4]: Do we need this line?  the refs name includes both branches and tags and is the source of a lot of trouble with non-reproducible downloads from github...

**Review by tacaswell — APPROVED** [review]
I will update the version of `setuptools_scm` we need in #23620

File: .git_archival.txt (hunk 2 above)
Comment (oscargus) [reply .git_archival.txt:4]: No idea. I just copied from https://github.com/Changaco/setuptools_scm_git_archive where they have that line in the migration guide.

Comment (tacaswell) [reply .git_archival.txt:4]: Lets defer this question to a later time.  It is a can of worms I do not want to block this moving forward.

**Pull request merged**